### PR TITLE
Formal Method, Type Theory: init

### PR DESCRIPTION
**Pull request by definfo**
Add Logic - Formal Method, Math/Logic Preliminaries - Type Theory.
Not sure if they are correctly classified.

Restrictions:
- Formal Method currently includes only Coq-related materials. For other proof assistants (Lean, Agda, ...) , you can check out the links in Type Theory part. (See [learn-tt](https://github.com/jozefg/learn-tt) / [∞-type Café Summer School](https://infinity-type-cafe.github.io/ntype-cafe-summer-school/) )
- Formal Method, Type Theory may also be labeled as "Programming Language Theory".

### CODE REVIEW

**Comment by mostafatouny** [comment]
Dear @definfo, Thanks for contributing.

>  For other proof assistants (Lean, Agda, ...) , you can check out the links in Type Theory part

Maybe you can add one more specific reference about Lean? We want only the best of the best.

> Formal Method, Type Theory may also be labeled as "Programming Language Theory".

I feel **Formal Methods** is the most basic foundational term. You can rigorously prove a software's correctness without getting close to _Programming Languages_

Remove [Infinity-Type-Cafe. ∞-type Café Summer School](https://infinity-type-cafe.github.io/ntype-cafe-summer-school/), As resources are aimed for a global audience. So English is mandatory.

Section _Math/Logic Preliminaries_ is for foundations needed by any theoretician. Any needs to know _proof by induction_ for example. So _Type Theory_ should never be included in that section.

Move [Daniel Gratzer. learn-type-theory](https://github.com/jozefg/learn-tt) to _Related Lists_ section.

Don't you think it's better to have:
- [nlabs](https://ncatlab.org/nlab/show/mathematics) - A curated list of mathematics including type theory and proof assistants, and more generally homotopy theory, algebra, topology, and category theory.
- [Software Foundations](https://softwarefoundations.cis.upenn.edu/) - A broad introduction series to the mathematical underpinnings of reliable software. It's composed of "proof scripts" for the Coq proof assistant. It's is intended for a broad range of readers, With no specific background assumed.

Since we are adding Type Theory, Then we need an accessible introduction to Functional Programming from the perspective of Formal Methods. What's your recommendation?

Please review and let me know your feedback of:
- [Formal Methods: An Appetizer](https://link.springer.com/book/10.1007/978-3-030-05156-3)
- [Cornell. Functional Programming in Ocaml](https://www.cs.cornell.edu/courses/cs3110/2024sp/)
- [ML for the working programmer](https://www.cl.cam.ac.uk/~lp15/MLbook/pub-details.html)

Please fix the unnecessary indentation of _Madhur Tulsiani. Mathematical Toolkit_ and _Harsha & Strivastava. Toolkit for Theoretical Computer Science. Tata Institute_ in the json file.

You Rock!

**Comment by definfo** [comment]
> I feel **Formal Methods** is the most basic foundational term. You can rigorously prove a software's correctness without getting close to _Programming Languages_

That's OK.
However, when it comes to real-world algorithm/software verification, the modelisation of program sometimes require us to specify a certain type system and its related properties, which is closely related to PL.

> Remove [Infinity-Type-Cafe. ∞-type Café Summer School](https://infinity-type-cafe.github.io/ntype-cafe-summer-school/), As resources are aimed for a global audience. So English is mandatory.

OK.

> Section _Math/Logic Preliminaries_ is for foundations needed by any theoretician. Any needs to know _proof by induction_ for example. So _Type Theory_ should never be included in that section.

You're right. I've missed the point.

> Move [Daniel Gratzer. learn-type-theory](https://github.com/jozefg/learn-tt) to _Related Lists_ section.
> 
> Don't you think it's better to have:
> 
>     * [nlabs](https://ncatlab.org/nlab/show/mathematics) - A curated list of mathematics including type theory and proof assistants, and more generally homotopy theory, algebra, topology, and category theory.
> 
>     * [Software Foundations](https://softwarefoundations.cis.upenn.edu/) - A broad introduction series to the mathematical underpinnings of reliable software. It's composed of "proof scripts" for the Coq proof assistant. It's is intended for a broad range of readers, With no specific background assumed.
> 
> 
> Since we are adding Type Theory, Then we need an accessible introduction to Functional Programming from the perspective of Formal Methods. What's your recommendation?

I prefer to put it seperately under *Programming Language* section, because FP is relatively independent and only a minority of FPers works in the field of Formal Verification.
Introduction to FP/TT has been taken into account by most Formal Method tutorials. Should be enough for beginners if they only care about the basic usage of proof assistant.

> Please review and let me know your feedback of:
> 
>     * [Formal Methods: An Appetizer](https://link.springer.com/book/10.1007/978-3-030-05156-3)

Seem that it focuses more on semantics analysis, runtime, etc.
This may be a material written in the aspect Information Security.

>     * [Cornell. Functional Programming in Ocaml](https://www.cs.cornell.edu/courses/cs3110/2024sp)

An excellent, comprehensive course of FP.
> If the best introductory course to programming is SICP. After that, it is CS3110.

>     * [ML for the working programmer](https://www.cl.cam.ac.uk/~lp15/MLbook/pub-details.html)
Also good, except that ML is outdated.

I would also recommend [Haskell MOOC](https://haskell.mooc.fi/).

> Please fix the unnecessary indentation of _Madhur Tulsiani. Mathematical Toolkit_ and _Harsha & Strivastava. Toolkit for Theoretical Computer Science. Tata Institute_ in the json file.

Sorry, I've encountered a display issue in VSCode. Will fix it in the latest push.

By the way, have you heard of the early practice of TCS classification into track A/B?
> Since that time M. Nivat, who is still Editor-in-Chief has reported regularly to council and general assembly and occasionally in the Bulletin - e.g. when the split into sections A (automata, algebra and algorithms) and B (logic, semantics and related topics) was decided upon (Bulletin no. 45, p.2,3, October 1991);

Thanks for review!

**Comment by mostafatouny** [comment]
> the modelisation of program sometimes require us to specify a certain type system and its related properties, which is closely related to PL

Our TCS Awesome List is targeted for a broad audience of CS and Math. When they read _Programming Language Theory_ what they will get is designing a new programming language. Maybe _Formal Software Verification_ is a better section name?

> Move [Daniel Gratzer. learn-type-theory](https://github.com/jozefg/learn-tt) to Related Lists section.

> I prefer to put it seperately under Programming Language section

This list is meant to be curated not complete. Readers should only see the best of the best resources for each field of Theoretical CS. Gratzer's list seems valuable but as it contains too many resources, Putting it in _Related Lists_ might be more useful to the reader.

> FP is relatively independent and only a minority of FPers works in the field of Formal Verification.

Okay. Let's separate Functional Programming from Formal Verification. Maybe we can do:
- Formal Software Verification
  - Functional Programming
  - Formal Methods / Proof Assistants / Theorem Provers / Programming Languages Theory (I haven't decided)

> Introduction to FP/TT has been taken into account by most Formal Method tutorials. Should be enough for beginners if they only care about the basic usage of proof assistant.

Do you recommend any accessible basic intro to both FP and TT?

> Seem that it focuses more on semantics analysis, runtime, etc. This may be a material written in the aspect Information Security.

Thanks for the feedback. I shall exclude it from consideration.

> If the best introductory course to programming is SICP.

Why is SICP considered to be the best? What makes it unique from others?

> Also good, except that ML is outdated.

Thanks. I shall exclude it from consideration.

> I would also recommend [Haskell MOOC](https://haskell.mooc.fi/).

Notably it has a community active on Telegram. A huge PLUS not available anywhere! Thanks, it deserves to be added.

> By the way, have you heard of the early practice of TCS classification into track A/B?

No I didn't. Maybe we can better organize the list. I am happy to listen from you.

**Comment by definfo** [comment]
> Our TCS Awesome List is targeted for a broad audience of CS and Math. When they read _Programming Language Theory_ what they will get is designing a new programming language. Maybe _Formal Software Verification_ is a better section name?

OK, I agree with you. I think we could just put it as *Formal Verification*, since this technique can be used not only for software, but also for algorithm (often appear as code snippet in lib files) or even for hardware (EDA).

> This list is meant to be curated not complete. Readers should only see the best of the best resources for each field of Theoretical CS. Gratzer's list seems valuable but as it contains too many resources, Putting it in _Related Lists_ might be more useful to the reader.

Got it. 

To better organize it, we could refer to [Programming language theory#Sub-disciplines and related fields](https://en.wikipedia.org/wiki/Programming_language_theory#Sub-disciplines_and_related_fields).

For example,

- Programming Language Theory
  - Formal Verification / Formal Methods
    - Proof Assistants (Coq/Lean/Agda/...)
  - Type theory
  - Programming Paradigm (TODO, see [Programming paradigm](https://en.wikipedia.org/wiki/Programming_paradigm))
    - Imperative
      - Procedural
      - Object-oriented
      - ...
    - Declarative
      - Functional (Move FP here)
      - ...
  - ...

This may not be the perfect solution but more readable and reasonable for me.

> Why is SICP considered to be the best? What makes it unique from others?

TLDR: SICP is a revolutionary course focusing on the central idea of *abstraction*, instead of tedious programming details.
You can check out [Why SICP matters](https://people.eecs.berkeley.edu/~bh/sicp.html).

**Comment by mostafatouny** [comment]
> put it as Formal Verification, since this technique can be used not only for software, but also for algorithm (often appear as code snippet in lib files) or even for hardware (EDA).

Okay

> For example,
Programming Language Theory
Formal Verification / Formal Methods
Proof Assistants (Coq/Lean/Agda/...)
Type theory
Programming Paradigm (TODO, see [Programming paradigm](https://en.wikipedia.org/wiki/Programming_paradigm))
Imperative
Procedural
Object-oriented
...
Declarative
Functional (Move FP here)
...
...

Okay. For now, We need _Formal Verification_, _Type Theory_, and _Functional Programming_. I suggest adding _Basics_ section with SICP also. So

- Programming Languages Theory
  - Basics
    - SICP
  - Formal Verification
  - Type Theory
  - Functional Programming

Feel free to update your branch, adding your resources following that organization. Please consider all aforementioned notes.

If anything is unclear, Feel free to ask.

File: big_list_builder/source.json [hunk 1]
```diff
@@ -299,8 +339,8 @@
         "TCS Inspired": {
             "Lecture Videos Playlists": [
                 "[O'Donnell. CS Theory Toolkit](https://www.youtube.com/playlist?list=PLm3J0oaFux3ZYpFLwwrlv_EHH9wtH6pnX) - It covers a large number of the math/CS topics that you need to know for reading and doing research in Computer Science Theory - alternatively: [bilibili](https://www.bilibili.com/video/BV1Ry4y1e7zR)",
-		"[Madhur Tulsiani. Mathematical Toolkit](https://home.ttic.edu/~madhurt/courses/toolkit2021/index.html) - Things prof. Madhur wish he knew in first year of grad school.",
-		"[Harsha & Strivastava. Toolkit for Theoretical Computer Science. Tata Institute](https://www.tifr.res.in/~prahladh/teaching/2020-21/toolkit/)"
+		    "[Madhur Tulsiani. Mathematical Toolkit](https://home.ttic.edu/~madhurt/courses/toolkit2021/index.html) - Things prof. Madhur wish he knew in first year of grad school.",
```
Comment (definfo) [inline big_list_builder/source.json:342]: I'm afraid that the indentation here is still problematic ...
